### DISCUSSION TITLE
Switch from a login window to the main window: best method?

### DISCUSSION THREAD

**Discussion by ShuhuaGao** [canton7/Stylet#260] [Q&A] [answered]
A common scenario is that a user has to first log in before seeing the main window of an application. I currently use the `WindowManager` to achieve this purpose but no sure whether it is the recommended way in Stylet. I demonstrate it with the toy project below.

- Project structure
![image](https://user-images.githubusercontent.com/20141984/121301946-82ca3a00-c92b-11eb-8a13-a1f17fdd624d.png)
Both `LoginView` and `MainView` inherit `Window` in XAML.
- Code highlight
  + Boostrapper.cs
  To launch `LoginView` first.
  ```cs
  public class Bootstrapper : Bootstrapper<LoginViewModel>
  ```
  + LoginViewModel.cs
   The window switching logic is placed in `Login`.
  ```cs
       public class LoginViewModel : Screen
    {
        private readonly IWindowManager windowManager;
        public LoginViewModel(IWindowManager windowManager)
        {
            this.windowManager = windowManager;
        }

        // check whether the user can login here
        public bool CanLogin()
        {
            return true;
        }

        // Action for the Login button
        public void Login()
        {
            if (CanLogin())
            {
                // show the main window
                var mainVM = new MainViewModel();
                windowManager.ShowWindow(mainVM);
                // close the current login window
                RequestClose();
            }
        }
    }
  ```

The complete project can be found [here](https://github.com/ShuhuaGao/WPF-Examples/tree/main/StyletPlayground/HelloLogin). Your review of the above code and any suggestions are appreciated.

**Comment by canton7** [comment, marked as answer]
The normal way is to have a Shell View/VM which just contains other views. It starts off showing the login view/VM, but then replaces that with the main view/VM. See the [wiki page on Screens and Conductors](https://github.com/canton7/Stylet/wiki/Screens-and-Conductors).

**Reply by ShuhuaGao** [reply]
Thanks. Is there any drawback of my approach? Since both `LoginView` and `MainView` are two `Window`s (rather than a `UserControl`) with possibly different window styles, I feel that `WindowManager` is more natural.

**Reply by canton7** [reply]
It's just how it appears to the user: your approach will flash between screens. That means you get the transitions where one's appearing and the other's disappearing, but it does mean they can be different sizes and have different window styles, etc. It's really down to what you want the user to experience.

**Reply by ShuhuaGao** [reply]
Thanks for your explanation. Got it.
